Concept: Default features / add default features section to troubleshooting guide (#255)
microsoft/vcpkg-docs: vcpkg/concepts/features.md

---
title: Understanding features in vcpkg
description: This article explains the concept of features, their capabilities, and their expected contents in detail.
author: JavierMatosD
ms.author: javiermat
ms.date: 02/21/2024
ms.topic: concept-article
---

# Concept: features

## Features

Manifests can define additive sets of functionality, behavior, and dependencies called _features_.

> Each feature should be additive with other features: if a project provides features `A` and `B`, then it should build successfully with both `A` and `B` enabled together. Features should not be used to define alternatives. For example, features cannot be used to choose between multiple, exclusive graphics APIs since it is not possible to choose both. Features can depend on other features within same project if the project's manifest has a [`"name"`](../reference/vcpkg-json.md#name).

Reviewer: What do you think about rewording this paragraph into something like this:

> By design, port features should follow these principles:
> 
> * Additive: enabling feature 'A' should provide new functionality otherwise missing from the package without disabling any other functionality.
> * Non-exclusive: enabling feature 'A' should not prevent feature 'B' from being installed.
>
> As such, features should not be used to define alternative sets of functionalities. For example, a graphics library should not use features to choose between exclusive graphics APIs, since it is not possible to choose both. 
>
> Features can have the following effects on a package's dependencies:
>
> * Add new dependencies, including dependencies on other features from the same package.
> * Enable new features on existing dependencies.

The set of features available are defined by the [`"features"` field](../reference/vcpkg-json.md#features).

### Example 1: Multiple file formats

An image manipulation library, for instance, might support several different image types by depending on different sets of other libraries.

```json
{
  "name": "my-image-lib",
  "version": "0.1",
  "features": {
    "png": { "description": "Support PNG files", "dependencies": ["libpng"]},
    "jpeg": { "description": "Support JPEG files", "dependencies": ["libjpeg-turbo"]},
    "tiff": { "description": "Support TIFF files", "dependencies": ["libtiff"]},
  }
}
```

## Default features

Default features are a set of features to be automatically activated if the top-level project does not explicitly request a build without them. Default features are intended to ensure a minimum level of functionality regardless of how complex and customizable the dependency graph of a project grows. 

>![NOTE]

> > They are not intended to model "curation" or "suggestions".

Reviewer: In my opinion, we should have a contributor guideline regarding default features.

---
title: Understanding features in vcpkg
description: This article explains the concept of features, their capabilities, and their expected contents in detail.
author: JavierMatosD
ms.author: javiermat
ms.date: 02/21/2024
ms.topic: concept-article
---

# Concept: features

## Features

> Manifests can define additive sets of functionality, behavior, and dependencies called _features_.

Reviewer: ```suggestion
Ports can define additive sets of functionality, behavior, and dependencies called _features_.
```

While they are defined in port manifests, we should avoid making it sound like a manifest exclusive feature.

> Manifests can define additive sets of functionality, behavior, and dependencies called _features_.

Reviewer: ```suggestion
Port _features_ represent sets of functionality, behavior, and dependencies that can be selectively added to a package upon installation. 
```

While they are defined in port manifests, we should avoid making it sound like a manifest exclusive feature.

Reviewer 2: Hm, with the suggested change, it sounds like a port exclusive feature. But top level manifests (aka apps) can use features, too.

---
title: Understanding features in vcpkg
description: This article explains the concept of features, their capabilities, and their expected contents in detail.
author: JavierMatosD
ms.author: javiermat
ms.date: 02/21/2024
ms.topic: concept-article
---

# Concept: features

## Features

Features represent sets of functionality, behavior, and dependencies that can be selectively added to a package or project upon installation. 

By design, features should follow these principles:

> - Additive: enabling feature A should provide new functionality otherwise missing from the package without disabling any other functionality.
> - Non-exclusive: enabling feature A should not prevent other features from being installed.

Reviewer: ```suggestion
- Additive: enabling a feature should provide new functionality otherwise missing from the package without disabling any other functionality.
- Non-exclusive: enabling a feature should not prevent other features from being installed.
```

After reading it again, I think indeterminate articles are appropriate here.

> As such, features should not be used to define alternative sets of functionalities. For example, a graphics library should not use features to choose between exclusive graphics backends, since it is not possible to install all of them at the same time.

Reviewer: ```suggestion
Features should not be used to define alternative sets of functionalities. For example, a graphics library should not use features to choose between exclusive graphics backends, since it's not possible to install all of them at the same time.
```

Features can have the following effects on a package's dependencies:

- Add new dependencies, including dependencies on other features from the same package.
- Enable new features on existing dependencies.

The set of features available are defined by the [`"features"` field](../reference/vcpkg-json.md#features).

### Example 1: Multiple file formats

An image manipulation library, for instance, might support several different image types by depending on different sets of other libraries.

```json
{
  "name": "my-image-lib",
  "version": "0.1",
  "features": {
    "png": { "description": "Support PNG files", "dependencies": ["libpng"]},
    "jpeg": { "description": "Support JPEG files", "dependencies": ["libjpeg-turbo"]},
    "tiff": { "description": "Support TIFF files", "dependencies": ["libtiff"]},
  }
}
```

## Default features

Default features are a set of features to be automatically activated if the top-level project does not explicitly request a build without them. Default features are intended to ensure a minimum level of functionality regardless of how complex and customizable the dependency graph of a project grows. 

>![NOTE]

> > They are not intended to model "curation" or "suggestions".

Reviewer: ```suggestion
> Default features are not intended to model "curation" or "suggestions".
```

For example, consider a library `"extract-any"` that supports over 10 different archive formats, including several that are quite obscure. Because they are all optional, if none are selected the library is not functional: it cannot extract any files.

> Default features ensure that a user who simply adds `"extract-any"` to the list of dependencies in their `vcpkg.json` will get a baseline level of functionality, for example automatically selecting `.zip` and `.tar.gz` decompressors.

Reviewer: ```suggestion
Default features ensure that a user who simply adds `"extract-any"` to the list of dependencies in their `vcpkg.json` will get a baseline level of functionality; for example, automatically selecting `.zip` and `.tar.gz` decompressors.
```

## Dependency resolution

When using vcpkg, dependency resolution plays a crucial role, especially when dealing with features that have interdependencies. To illustrate, consider the following scenario involving an image manipulation library:

```json
{
  "name": "my-image-lib",
  "version": "0.1",
  "features": {
    "png": { "description": "Support PNG files", "dependencies": ["libpng"]},
    "jpeg": { "description": "Support JPEG files", "dependencies": ["libjpeg-turbo"]},
    "tiff": { "description": "Support TIFF files", "dependencies": ["libtiff"]},
  }
}
```

In scenarios where different libraries depend on various features of a common library, vcpkg ensures that all required features and dependencies are considered. For instance, if `library-a` requires the `png` feature and `library-b` requires the `jpeg` feature from `my-image-lib`, the dependency graph would look like this:

```json
{
  "name": "library-a",
  "version": "1",
  "dependencies": [{"name": "my-image-lib", "features": ["png"]}]
}
```

```json
{
  "name": "library-b",
  "version": "1",
  "dependencies": [{"name": "my-image-lib", "features": ["jpeg"]}]
}

> ```

Reviewer: Just for completion, include a third manifest that declares dependencies in `library-a` and `library-b`.
